File: seminar/src/main/java/com/server/dosopt/seminar/domain/Member.java
package com.server.dosopt.seminar.domain;

import jakarta.persistence.*;   // java 17버전 이상부터는 jakarta로 변경
import lombok.AccessLevel;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)   // 엔티티에는 반드시 고유 식별자(ID)가 존재해야 핢 -> 생성전략을 함께 지정
    private Long id;
    private String name;
    private String nickname;
    private int age;

    @Embedded   // 값 타입을 사용하는 곳에 붙이는 어노테이션
    private SOPT sopt;

    @Builder   // 빌더패턴을 적용하여 객체 생성 가능
    public Member(String name, String nickname, int age, SOPT sopt) {
        this.name = name;
        this.nickname = nickname;
        this.age = age;
        this.sopt = sopt;
    }
}

File: kotlin-seminar/src/main/kotlin/sopt/org/kotlinseminar/domain/Member.kt
package sopt.org.kotlinseminar.domain

import jakarta.persistence.Column
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id

@Entity
class Member(
    name: String,

    nickname: String,

    age: Int,

    sopt: Sopt,

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0L
): BaseTimeEntity() {

    @Column
    var name = name
        protected set

    @Column
    var nickname = nickname
        protected set   // setter를 외부에서 호출하지 못하도록 막아둔다

    @Column
    var age = age
        protected set

    @Embedded
    var sopt = sopt
        protected set
}